src/types/widget.ts

```diff
@@ -300,7 +284,73 @@ export type Widget = {
     state: SearchParameters,
     widgetSearchParametersOptions: WidgetSearchParametersOptions
   ): SearchParameters;
-};
+} & (
```

Comment:
With this strategy, custom widgets that implement `getWidgetRenderState` will throw a TypeScript error. We still need to add a generic `getWidgetRenderState` independent of the `$$type`.

Reply:
That's a good point.

Reply:
Yes, and I don't see why this is an issue :thinking:

Reply:
IIUC, if a user is using TypeScript, and implement a custom widget with a $$type: let's say.. `ais.myCustomWidget`. Then the custom widget's type `Widget` won't include `getWidgetRenderState` and the user has no proper way to include it.

Reply:
What I mean is that we can (must) add a generic `getWidgetRenderState` method to the base Widget type to avoid this and for the user to get proper typing in a custom widget.

Also users must not put a `$$type` key on a custom widget since this is for internal use only.